apps/webapp/src/state/staking.ts

```diff
@@ -219,6 +220,43 @@ export const createStakingSlice = (): SliceCreator<StakingSlice> => (set, get) =
       }
     }
   },
+  undelegate: async () => {
+    const toast = new TransactionToast('undelegate');
+    toast.onStart();
+
+    try {
+      const transactionPlan = await plan(assembleUndelegateRequest(get().staking));
+
+      // Reset form _after_ building the transaction form, since it depends on
+      // the state.
+      set(state => {
+        state.staking.amount = '';
+        state.staking.action = undefined;
+        state.staking.validatorInfo = undefined;
+      });
```

Comment:
If this throws an error, the user has to re-enter the data. Do we want this? If not, think this could go in the `finally` block.

Reply:
Good call — I reset the `amount` in the `finally` block, but left the `action`/`validatorInfo` here at the top, so that the form dialog closes as soon as they submit it.